pkgs/development/compilers/tinygo/default.nix

```
@@ -1,64 +1,74 @@
-{ lib, buildGoModule, fetchFromGitHub, llvm, clang-unwrapped, lld, avrgcc
-, avrdude, openocd, gcc-arm-embedded, makeWrapper, fetchurl }:
+{ lib
+, buildGoModule
+, fetchFromGitHub
+, makeWrapper
+, llvm
+, clang-unwrapped
+, lld
+, avrgcc
+, go
+, gcc-arm-embedded
+, avrdude
+, openocd
+}:
 
-let main = ./main.go;
-    gomod = ./go.mod;
-in
-buildGoModule rec {
+let
+  llvmMajor = lib.versions.major llvm.version;
+in buildGoModule rec {
   pname = "tinygo";
-  version = "0.16.0";
+  version = "0.21.0";
 
   src = fetchFromGitHub {
     owner = "tinygo-org";
-    repo = "tinygo";
+    repo = pname;
     rev = "v${version}";
-    sha256 = "063aszbsnr0myq56kms1slmrfs7m4nmg0zgh2p66lxdsifrfly7j";
+    sha256 = "sha256-ZAUvOKApwOIKWINta9jEs/M32pHgW86+a0Y1jHEMupk=";
     fetchSubmodules = true;
   };
 
-  overrideModAttrs = (_: {
-      patches = [];
-      preBuild = ''
-      rm -rf *
-      cp ${main} main.go
-      cp ${gomod} go.mod
-      chmod +w go.mod
-      '';
-  });
+  vendorSha256 = "sha256-fD9/hPRXIT7cPljA+QNWDRESqHXeCbtWHMq12280R1g=";
 
-  preBuild = "cp ${gomod} go.mod";
-
-  postBuild = "make gen-device";
+  nativeBuildInputs = [ makeWrapper ];
+  buildInputs = [ llvm clang-unwrapped ];
 
-  vendorSha256 = "12k2gin0v7aqz5543m12yhifc0xsz26qyqra5l4c68xizvzcvkxb";
+  subPackages = [ "." ];
 
   doCheck = false;
 
-  prePatch = ''
+  # We need to set GOROOT to import and compile
+  # standard go libraries with tinygo
+  allowGoReference = true;
+
+  postPatch = ''
+    patchShebangs lib/wasi-libc
```

I don't know what `patchShebangs` does, but from the name, it might not be necessary.
TinyGo does not run shell scripts itself, although shell scripts do need to be run when building wasi-libc (which must be done while building the tinygo package).

---

It patches the shebang of https://github.com/WebAssembly/wasi-libc/blob/ad5133410f66b93a2381db5b542aad5e0964db96/libc-top-half/musl/tools/install.sh to the path of bash in the nix store.

also the path should probably be

```suggestion
    patchShebangs lib/wasi-libc/libc-top-half/musl/tools/install.sh
```

---

I suspect we don't use that file. Certainly not at runtime (see https://github.com/NixOS/nixpkgs/pull/141212#discussion_r754227119) but probably not at build time either. Patching it won't hurt either so :shrug: